File: course-info/fetch-exam.py
import requests
import pytz
import json
from typing import List
from datetime import datetime
from bs4 import BeautifulSoup

local_tz = pytz.timezone('America/New_York')
now = datetime.now()


class ExamInfo(object):
    subject: str
    course_number: str
    section_number: str
    time: str

    def __init__(self, subject: str, course_number: str,
                 section_number: str, time: str):
        self.subject = subject
        self.course_number = course_number
        self.section_number = section_number
        self.time = time

    def to_dict(self) -> dict:
        return dict(
            subject=self.subject, courseNumber=self.course_number,
            sectionNumber=self.section_number, time=self.time
        )

    @classmethod
    def parse_prelim_line(cls, line: str) -> 'ExamInfo':
        segments = line.split(sep=' ')
        datetime_str = ''
        for i in range(2, len(segments)):
            s = segments[i]
            if s == '':
                continue
            else:
                datetime_str = s
                break
        datetime_str += ' 7:30 PM'

        datetime_without_tz = datetime.strptime(
            datetime_str, '%m/%d/%Y %I:%M %p')
        datetime_with_tz = local_tz.localize(datetime_without_tz)
        datetime_in_utc: datetime = datetime_with_tz.astimezone(pytz.utc)

        return ExamInfo(
            subject=segments[0],
            course_number=segments[1],
            section_number='',
            time=datetime_in_utc.__str__().split('+')[0]
        )

    @classmethod
    def parse_final_line(cls, line: str) -> 'ExamInfo':
        segments = line.split(sep=' ')
        datetime_str = f'{now.year}'
        for i in range(3, len(segments)):
            s = segments[i]
            if s == '':
                continue
            datetime_str += ' '
            datetime_str += s
        datetime_str = datetime_str.strip()

        datetime_without_tz = datetime.strptime(
            datetime_str, '%Y %a, %b %d %I:%M %p')
        datetime_with_tz = local_tz.localize(datetime_without_tz)
        datetime_in_utc: datetime = datetime_with_tz.astimezone(pytz.utc)

        return ExamInfo(
            subject=segments[0],
            course_number=segments[1],
            section_number=segments[2],
            time=datetime_in_utc.__str__().split('+')[0]
        )


def get_exam_block(url: str) -> str:
    return BeautifulSoup(requests.get(url=url).text, 'html.parser').pre.string


def get_exam_info_list(url: str, is_final: bool) -> List[ExamInfo]:
    exam_block = get_exam_block(url=url)
    lines = exam_block.splitlines()
    info_list = []
    for line in lines:
        if line == '' or line.startswith('Class'):
            continue
        info = ExamInfo.parse_final_line(line) \
            if is_final \
            else ExamInfo.parse_prelim_line(line)
        info_list.append(info)
    return info_list


def create_json(url: str, is_final: bool, filename: str):
    exam_info_list = [
        info.to_dict()
        for info in get_exam_info_list(url=url, is_final=is_final)
    ]
    with open(filename, 'w') as outfile:
        json.dump(exam_info_list, outfile, indent=2, sort_keys=True)


if __name__ == '__main__':
    create_json(
        url='https://registrar.cornell.edu/exams/spring-final-exam-schedule',
        is_final=True,
        filename='final-exams.json'
    )
    create_json(
        url='https://registrar.cornell.edu/exams/spring-prelim-schedule',
        is_final=False,
        filename='prelim-exams.json'
    )

File: course-info/fetch-exam.js
const fs = require('fs');
const fetch = require('node-fetch');
const { JSDOM } = require('jsdom');

const fallPrelimUrl = 'https://registrar.cornell.edu/exams/fall-prelim-exam-schedule';
const fallFinalUrl = 'https://registrar.cornell.edu/exams/fall-final-exam-schedule';

const currentYear = new Date().getFullYear();

async function fetchExamText(url) {
  const resp = await fetch(url);
  const html = await resp.text();
  const dom = new JSDOM(html);
  const preNode = dom.window.document.querySelector('pre');
  return preNode.innerHTML;
}

function parsePrelimLine(line) {
  const segments = line.split(' ');
  const subject = segments[0];
  const courseNumber = segments[1];
  let datetimeString = '';
  for (let i = 2; i < segments.length; i += 1) {
    const s = segments[i];
    if (s !== '') {
      datetimeString = s;
      break;
    }
  }
  const date = new Date(datetimeString);
  date.setFullYear(currentYear);
  date.setHours(19, 30); // 7:30 prelim time
  return {
    subject,
    courseNumber,
    sectionNumber: '',
    time: date.getTime(),
  };
}

function parseFinalLine(line) {
  const segments = line.split(' ').filter((part) => part !== '');
  const subject = segments[0];
  const courseNumber = segments[1];
  const sectionNumber = segments[2];
  let datetimeString = '';
  for (let i = 3; i < segments.length; i += 1) {
    const s = segments[i];
    if (s === 'final') {
      break;
    }
    if (s !== '') {
      datetimeString += ` ${s}`;
    }
  }
  datetimeString = datetimeString.trim();
  const date = new Date(datetimeString);
  date.setFullYear(currentYear);
  return {
    subject,
    courseNumber,
    sectionNumber,
    time: date.getTime(),
  };
}

function getExamInfoList(rawText, isFinal) {
  const lines = rawText.split('\n');
  const infoList = [];
  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i].trim();
    if (line !== '' && !line.startsWith('Class')) {
      const info = isFinal ? parseFinalLine(line) : parsePrelimLine(line);
      infoList.push(info);
    }
  }
  return infoList;
}

function createJson(url, isFinal, outFilename) {
  fetchExamText(url)
    .then((rawText) => getExamInfoList(rawText, isFinal))
    .then((json) => fs.writeFile(outFilename, JSON.stringify(json), () => {}));
}

function main() {
  createJson(fallFinalUrl, true, 'final-exams.json');
  createJson(fallPrelimUrl, false, 'prelim-exams.json');
}

main();
